include/up-cpp/utils/SafeMap.h

```diff
@@ -0,0 +1,323 @@
+        std::unique_lock lock(map_lock_);
+        return f(*this);
+    }
+
+    using ConstTxn = std::function<std::any(MapT const&)>;
+    std::any transact(ConstTxn &&f) const {
+        std::shared_lock lock(map_lock_);
+        return f(*this);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    ///////////////////// Pass-throughs for std::map //////////////////////////
+    ///////////////////////////////////////////////////////////////////////////
+    // Pass through all the type definitions (except those specific to
+    // situations where iterators can / must be used.
+    using key_type = typename MapT::key_type;
+    using mapped_type = typename MapT::mapped_type;
+    using value_type = typename MapT::value_type;
+    using size_type = typename MapT::size_type;
+    using difference_type = typename MapT::difference_type;
+    //using key_compare = typename MapT::key_compare;
+    //using value_compare = typename MapT::value_compare;
+    using allocator_type = typename MapT::allocator_type;
+    using reference = typename MapT::reference;
+    using const_reference = typename MapT::const_reference;
+    using pointer = typename MapT::pointer;
+    using const_pointer = typename MapT::const_pointer;
+    //using iterator = typename MapT::iterator;
+    //using const_iterator = typename MapT::const_iterator;
+    //using reverse_iterator = typename MapT::reverse_iterator;
+    //using const_reverse_iterator = typename MapT::const_reverse_iterator;
+    using node_type = typename MapT::node_type;
+    //using insert_return_type = typename MapT::insert_return_type;
+    ////TODO Base-class-specific using statements (e.g. hasher)
+
+    // Pass through the base map class constructors to allow construction
+    // from an unsafe map. Also covers some of the generic iterator
+    // constructors.
+    using MapT::MapT;
+
+    // Define the constructors we need for copy and move
+    BaseSafeMap() = default;
+
+    BaseSafeMap(const SafeMapT &other)
+    : CMCLock(LockMode::Shared, other.map_lock_), MapT(other) {
+        CMCLock::unlock();
+    }
+
+    BaseSafeMap(const SafeMapT &other, const allocator_type &alloc)
+    : CMCLock(LockMode::Shared, other.map_lock_), MapT(other, alloc) {
+        CMCLock::unlock();
+    }
+
+    BaseSafeMap(SafeMapT &&other)
+    : CMCLock(LockMode::Unique, other.map_lock_), MapT(std::move(other)) {
+        CMCLock::unlock();
+    }
+
+    BaseSafeMap(SafeMapT &&other, const allocator_type &alloc)
+    : CMCLock(LockMode::Unique, other.map_lock_), MapT(std::move(other), alloc) {
+        CMCLock::unlock();
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    /////////// Locking wrappers for existing std::map interfaces /////////////
+    ///////////////////////////////////////////////////////////////////////////
+    SafeMapT& operator=(const SafeMapT &other) {
+        std::scoped_lock lock(map_lock_, other.map_lock_);
```

Comment:
I just noticed that this is probably wrong - `scoped_lock` will be creating a unique lock on both mutexes. However, only a shared_lock should be applied to other in this location. I will need to look if there is an equivalent to scoped_lock that allows for mixed locking modes like that.